dhis-2/dhis-services/dhis-service-validation/src/test/java/org/hisp/dhis/validation/ValidationResultStoreTest.java

```diff
@@ -490,126 +471,312 @@ public void testGetById() throws Exception
+            validationResultStore.query( query ) );
+
+        // filter on B gives results for B
+        query.setVr( singletonList( validationRuleB.getUid() ) );
+        assertEqualSets( asList( validationResultBA, validationResultBB, validationResultBC ),
+            validationResultStore.query( query ) );
+
+        // case with multiple units
+        query.setVr( asList( validationRuleA.getUid(), validationRuleB.getUid() ) );
+        assertEqualSets( asList(
+            validationResultAA, validationResultAB, validationResultAC,
+            validationResultBA, validationResultBB, validationResultBC ),
+            validationResultStore.query( query ) );
+
+        // now we restrict user to only be able to see Bs
+        setMockUserService( userBService );
+        // so filtering on As should not give any result
+        query.setVr( singletonList( validationRuleA.getUid() ) );
+        assertEqualSets( emptyList(), validationResultStore.query( query ) );
+    }
+
+    @Test
+    public void testQueryWithIsoPeriodFilter() {
+        save( asList(
+            validationResultAA, validationResultAB, validationResultAC,
+            validationResultBA, validationResultBB, validationResultBC ) );
+
+        // test with superuser so user adds no extra restrictions
+        setMockUserService( superUserService );
+
+        ValidationResultQuery query = new ValidationResultQuery();
+
+        // periodA is Jan 2017, periodB is Feb 2017
+        // monthly ISO pattern: YYYY-MM
+        query.setPe( singletonList( "2017-01" ) );
+        assertEqualSets( asList( validationResultAA, validationResultAB, validationResultAC ),
+            validationResultStore.query( query ) );
+
+        query.setPe( asList( "2017-01", "2017-02" ) );
+        assertEqualSets( asList(
+            validationResultAA, validationResultAB, validationResultAC,
+            validationResultBA, validationResultBB, validationResultBC),
+            validationResultStore.query( query ) );
+
+        // QUARTERLY
+        query.setPe( singletonList( "2017Q1" ) );
+        assertEqualSets( asList(
+            validationResultAA, validationResultAB, validationResultAC,
+            validationResultBA, validationResultBB, validationResultBC),
+            validationResultStore.query( query ) );
+
+        // YEARLY
+        query.setPe( singletonList( "2017" ) );
+        assertEqualSets( asList(
+            validationResultAA, validationResultAB, validationResultAC,
+            validationResultBA, validationResultBB, validationResultBC),
+            validationResultStore.query( query ) );
+
+        // WEEKLY
+        query.setPe( singletonList( "2017W3" ) );
+        assertEqualSets( asList( validationResultAA, validationResultAB, validationResultAC ),
+            validationResultStore.query( query ) );
+    }
+
+    @Test
+    public void testQueryWithCreatedDateFilter() {
+        Date beforeA = new Date();
+        wait1();
```

Comment:
I think we can remove this test to avoid waiting/timing, will slow down the tests and might be unstable on the CI server.

Reply:
It is waiting 1ms - do you really think we don't have that time?

Reply:
changed the name to `wait1ms` to make this clear.